README.md
Edge (Ruler for Chrome)
=======================

Edge is a simple on-screen ruler written using the Chrome packaged app platform. 


Installation
------------

#### Most users will want to install [from the Chrome Web Store][webstore-dl].

### Installing from the source

* Download [this zip of Edge][edge-dl]
* Unzip Ruler-master.zip
* On chrome://extensions click the "Enable Developer Mode" checkbox
* Click the "Load unpacked extension..." button and select the app subdirectory of Edge
* Run the app from the new tab page

Technologies
------------

* [Chrome App Platform][chrome-app-platform]
* [AngularJS][angularjs]

Authors
-------

* Andrew Stromme &lt;andrew.stromme@gmail.com&gt;


[edge-dl]: https://github.com/astromme/Ruler/archive/master.zip
[webstore-dl]: https://chrome.google.com/webstore/detail/njlkegdphefeellhaongiopcfgcinikh?hl=en&gl=US
[chrome-app-platform]: http://developer.chrome.com/stable/apps/about_apps.html
[angularjs]: http://angularjs.org/


CHANGELOG.md
Edge Changelog
==============

v0.6.6.2
--------

* fix ruler for Chrome 38

v0.6.6.1
--------

* fix vertical ruler by having a fallback location for always_on_top state

v0.6.6
------

* add nudging width/height with alt key
* better release tooling
* old releases now stored in git repo

v0.6.5
------

* add always on top feature

v0.6.4.1
--------

* accidentally included work-in-progress code when packaging up 0.6.4. No other changes

v0.6.4
------

* Fix a race condition causing rulers opened at start to be blank

v0.6.3
------

* Feature: Remember position & units of rulers when starting up
* Bugfix: Limit screen diagonal value to 999 to prevent unusable behavior
* Bugfix: Don't move ruler when arrow keys are used inside an input

v0.6.2
------

* Another small branding change

v0.6.1
------

* Small branding change
* Arrow keys move the ruler around

v0.6
----

* Name change to Edge

v0.5.2
------

* More analytics
* Improved usability (see below)
* Expose all 3 types of units all of the time
* Expose the calibration all of the time on the relevant units
* Removed blue automatic outlines from buttons

v0.5.1
------

* Fix Analytics support

v0.5
----

* Support for entering screen inches/centimeters to calibrate
* Fix Analytics support for vertical rulers

v0.4
----

* Google Analytics support

v0.3.1
------

* Fix creating new horizontal/vertical rulers on Win/Mac/OSX

v0.3
----

* Support for vertical rulers!
* Create more horizontal/vertical rulers by clicking the buttons near the units label

v0.2.3
------

* Update description

v0.2.2
------

* Fix changing between px, cm, in on Chrome OS and Windows
* Fix drawing bug when used with Chrome 30

v0.2.1
------

* New icon

v0.2
---

* Support for changing units between px, cm, in
* Show units label
* Massive performance improvement
* Support for resizing ruler
* Close button is a little nicer

v0.1.2
-----

* fix close button on Windows, CrOS. Issue #8

v0.1.1
-----

* Prevent resizing to fix redrawing bug

v0.1
---

* First released version
* Ruler supports px units

app/background.js
if (typeof String.prototype.startsWith != 'function') {
  String.prototype.startsWith = function (str){
    return this.slice(0, str.length) == str;
  };
}

if (typeof String.prototype.endsWith != 'function') {
  String.prototype.endsWith = function (str){
    return this.slice(-str.length) == str;
  };
}

function generateGUID() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
        return v.toString(16);
    });
}

var base_parameters = {
  'frame' : 'none',
}

var horizontal_limits = {
  'minWidth':300,
  'minHeight':50,
  'maxWidth':4000,
  'maxHeight':50,
}

var vertical_limits = {
  'minWidth':50,
  'minHeight':300,
  'maxWidth':50,
  'maxHeight':4000,
}

function win_create_callback(win) {
  win.onBoundsChanged.addListener(function() {
    console.log("onBoundsChanged()");
  });
}

function createHorizontalRuler() {
  var parameters = {
    'width':700,
    'height':50,
    'id' : generateGUID()
  }

  for (key in base_parameters) {
    parameters[key] = base_parameters[key];
  }

  for (key in horizontal_limits) {
    parameters[key] = horizontal_limits[key];
  }

  chrome.app.window.create('ruler_horizontal.html', parameters, win_create_callback);
}


function createVerticalRuler() {
  var parameters = {
    'width':50,
    'height':500,
    'frame' : 'none',
    'id' : generateGUID()
  }

  for (key in base_parameters) {
    parameters[key] = base_parameters[key];
  }

  for (key in vertical_limits) {
    parameters[key] = vertical_limits[key];
  }

  chrome.app.window.create('ruler_vertical.html', parameters, win_create_callback);
}

function recreateExistingWindow(id, orientation) {
  console.log('recreating' + id + orientation);

  var parameters = {
    'frame' : 'none',
    'id' : id
  };
  
  for (key in base_parameters) {
    parameters[key] = base_parameters[key];
  }

  var limits = orientation == 'horizontal' ? horizontal_limits : vertical_limits;
  for (key in limits) {
    parameters[key] = limits[key];
  }

  chrome.app.window.create(orientation == 'horizontal' ?
      'ruler_horizontal.html' : 'ruler_vertical.html', parameters, win_create_callback);
}

chrome.app.runtime.onLaunched.addListener(function() {
  var ruler_created = false;

  chrome.storage.local.get(null, function(data) {
    console.log(data);
    for (var key in data) {
      if (key.startsWith('ruler-')) {
        recreateExistingWindow(key.slice('ruler-'.length),
                               data[key].orientation);
        ruler_created = true;
      }
    }

    if (!ruler_created) {
      createHorizontalRuler();
    }
  });
});

chrome.runtime.onMessage.addListener(function(message,sender,sendResponse){
  if (message.text == "createHorizontalRuler") {
    createHorizontalRuler();
  } else if (message.text == "createVerticalRuler") {
    createVerticalRuler();
  }
});

chrome.app.window.onBoundsChanged.addListener(function() {
  console.log("onBoundsChanged()");
  console.log(arguments);
});

app/app.js
var ruler_app = angular.module('ruler', []);

// ruler can't get bigger than this
var maximum_width = "4000"

// 2.54 cms in one inch
var cm_per_in = 2.54;

// keycodes
var left_arrow = 37;
var up_arrow = 38;
var right_arrow = 39;
var down_arrow = 40;

var service = analytics.getService('ruler');
var tracker = service.getTracker('UA-40067294-2');  // Supply your GA Tracking ID.

function toArray(list) {
  return Array.prototype.slice.call(list || [], 0);
}

Array.prototype.remove = function() {
    var what, a = arguments, L = a.length, ax;
    while (L && this.length) {
        what = a[--L];
        while ((ax = this.indexOf(what)) !== -1) {
            this.splice(ax, 1);
        }
    }
    return this;
};

function isFunction(functionToCheck) {
 var getType = {};
 return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';
}


// called when the pointer lock has changed. Here we check whether the
// pointerlock was initiated on the element we want.
function changeCallback(e) {
    if (document.webkitPointerLockElement === document.getElementById("body")) {

        // we've got a pointerlock for our element, add a mouselistener
        document.addEventListener("mousemove", moveCallback, false);
    } else {

        // pointer lock is no longer active, remove the callback
        document.removeEventListener("mousemove", moveCallback, false);

        // and reset the entry coordinates
        entryCoordinates = {x:-1, y:-1};
    }
};

function errorHandler(e) {
  console.error(e);
}

ruler_app.controller('RulerControl', ['$scope', function($scope) {
    $scope.horizontal = {
        units: null,
        units_label: null,
        subdivisions_per_unit: null,
        tick_spacing: null,
        minor_per_major: null,
        minor_per_label: null,

        unit_options: {
            "px": {
                units: "px",
                units_label: "px",
                subdivisions_per_unit: 1,
                tick_spacing: 2,
                minor_per_major: 5,
                minor_per_label: 25,
            },
            "cm": {
                units: "cm",
                units_label: "cm",
                subdivisions_per_unit: 10,
                tick_spacing: 1,
                minor_per_major: 5,
                minor_per_label: 10,
            },
            "in": {
                units: "in",
                units_label: "in",
                subdivisions_per_unit: 8,
                tick_spacing: 1,
                minor_per_major: 4,
                minor_per_label: 8,
            },
        }
    }

    $scope.settings_visible = false;

    $scope.screen_diagonal_inches = 0;
    $scope.screen_diagonal_cm = 0;

    //console.log("adding message listener");
    chrome.runtime.onMessage.addListener(function(message,sender,sendResponse){
        //console.log("got message", message)
        if (message.text == "updateScreenInches") {
            $scope.screen_diagonal_inches = message.inches;
            $scope.updateScreenInches(true /*suppress_broadcast*/);
        } else if (message.text == "updateScreenCm") {
            $scope.screen_diagonal_cm = message.cm;
            $scope.updateScreenCm(true /*suppress_broadcast*/);
        }
    });

    $scope.closeWindow = function() {
        chrome.storage.local.remove('ruler-'+chrome.app.window.current().id);
        tracker.sendEvent('Ruler', 'CloseClicked');
        tracker.sendAppView('Ruler-'+$scope.orientation);
        window.close();
    }

    $scope.showSettings = function() {
        if ($scope.settings_visible) {
            tracker.sendEvent('Ruler', 'ShowSettings');
        } else {
            tracker.sendEvent('Ruler', 'HideSettings');
        }

        $scope.settings_visible = !$scope.settings_visible;
    }

    $scope.isTypeOfUnit = function(unit) {
        return $scope.horizontal.units == unit;
    }

    $scope.isPx = function() { return $scope.isTypeOfUnit('px'); }
    $scope.isCm = function() { return $scope.isTypeOfUnit('cm'); }
    $scope.isIn = function() { return $scope.isTypeOfUnit('in'); }

    $scope.selectTypeOfUnit = function(unit) {
        tracker.sendAppView('Ruler-'+$scope.orientation+'-'+unit);
        tracker.sendEvent('Ruler', 'UnitsClicked');

        $scope.selectUnits(unit);
    }

    $scope.selectPx = function() { return $scope.selectTypeOfUnit('px'); }
    $scope.selectCm = function() { return $scope.selectTypeOfUnit('cm'); }
    $scope.selectIn = function() { return $scope.selectTypeOfUnit('in'); }

    $scope.updateScreenInches = function(suppress_broadcast) {
        $scope.screen_diagonal_cm = cm_per_in*$scope.screen_diagonal_inches;
        $scope.selectUnits($scope.horizontal.units);

        if (suppress_broadcast == true) {
            //console.log("Suppressing broadcast");
            return;
        }

        //console.log("Telling other rulers to update their screen inches");
        chrome.runtime.sendMessage({ text:"updateScreenInches",
                                     inches:$scope.screen_diagonal_inches },
                                   function(reponse){});
        $scope.saveSettings(); 
    }

    $scope.updateScreenCm = function(suppress_broadcast) {
        $scope.screen_diagonal_inches = $scope.screen_diagonal_cm/cm_per_in;
        $scope.selectUnits($scope.horizontal.units);

        if (suppress_broadcast == true) {
            //console.log("Suppressing broadcast");
            return;
        }

        //console.log("Telling other rulers to update their screen inches");
        chrome.runtime.sendMessage({ text:"updateScreenCm",
                                     cm:$scope.screen_diagonal_cm },
                                   function(reponse){});
        $scope.saveSettings();
    }  

    $scope.always_on_top_state = function() {
        var checkbox = document.getElementById('always_on_top');
        if (checkbox) {
            return checkbox.checked;
        } else {
            return $scope.always_on_top;
        }
    }

    $scope.set_always_on_top_state = function(state) {
        if (state == undefined) {
            state = false;
        }
        
        var checkbox = document.getElementById('always_on_top');
        if (checkbox) {
            checkbox.checked = state;
        } else {
            $scope.always_on_top = state;
        }

        chrome.app.window.current().setAlwaysOnTop(state);
    }

    $scope.saveSettings = function(callback) {
        var data = {
            'screen_diagonal_inches': $scope.screen_diagonal_inches,
            'always_on_top': $scope.always_on_top_state()
        }

        var id = 'ruler-'+chrome.app.window.current().id;
        data[id] = {
            'orientation': $scope.orientation,
            'units': $scope.horizontal.units
        };

        chrome.storage.local.set(data, function() {
            console.log("settings saved");
            if (isFunction(callback)) {
                callback();
            }
        });
    }

    $scope.loadSettings = function(callback) {
        chrome.storage.local.get(null, function(data) {
            $scope.set_always_on_top_state(data.always_on_top);

            $scope.screen_diagonal_inches = data.screen_diagonal_inches;
            $scope.screen_diagonal_cm = cm_per_in*$scope.screen_diagonal_inches;

            var id = 'ruler-'+chrome.app.window.current().id;
            if (data[id]) {
                $scope.selectUnits(data[id].units)
            } else {
                $scope.selectUnits("px");
                $scope.saveSettings();
            }

            console.log("settings loaded");
            if (isFunction(callback)) {
                callback();
            }
        });
    }

    $scope.selectUnits = function(units) {
        var unit_options = $scope.horizontal.unit_options[units];
        for (var key in unit_options) {
            $scope.horizontal[key] = unit_options[key];
        }

        $scope.saveSettings();

        var ctx = document.getElementById("ruler_canvas").getContext("2d");
        ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);

        $scope.draw_ruler();
    }

    $scope.keydown = function($event) {
        if ($event.target.tagName == "INPUT") {
            return; // ignore custom keydown for input events
        }

        var bounds = chrome.app.window.current().getBounds();

        if ($event.keyCode == left_arrow) {
            if ($event.altKey && $scope.orientation == 'horizontal') {
                bounds.width -= $event.shiftKey ? 100 : 10;
            } else {
                bounds.left -= $event.shiftKey ? 10 : 1;
            }
        } else if ($event.keyCode == up_arrow) {
            if ($event.altKey && $scope.orientation == 'vertical') {
                bounds.height -= $event.shiftKey ? 100 : 10;
            } else {
                bounds.top -= $event.shiftKey ? 10 : 1;
            }
        } else if ($event.keyCode == right_arrow) {
            if ($event.altKey && $scope.orientation == 'horizontal') {
                bounds.width += $event.shiftKey ? 100 : 10;
            } else {
                bounds.left += $event.shiftKey ? 10 : 1;
            }
        } else if ($event.keyCode == down_arrow) {
            if ($event.altKey && $scope.orientation == 'vertical') {
                bounds.height += $event.shiftKey ? 100 : 10;
            } else {
                bounds.top += $event.shiftKey ? 10 : 1;
            }
        }

        chrome.app.window.current().setBounds(bounds);
    }

    $scope.mousemove = function($event) {
        //console.log("mousemove");
        console.log("press: " + $event.button);
        //console.log($event.webkitMovementX + " " + $event.webkitMovementY);
        //console.log(window.screenX + " " + window.screenY);
    }

    $scope.mousemove = function($event) {
        //console.log("mousemove");
        console.log($event.button);
        //console.log($event.webkitMovementX + " " + $event.webkitMovementY);
        //console.log(window.screenX + " " + window.screenY);
    }

    $scope.createHorizontalRuler = function($event) {
        tracker.sendEvent('Ruler', 'CreateHorizontalRuler', 'from-'+$scope.orientation);

        chrome.runtime.sendMessage({text:"createHorizontalRuler"}, function(reponse){
        });
    }

    $scope.createVerticalRuler = function($event) {
        tracker.sendEvent('Ruler', 'CreateHorizontalRuler', 'from-'+$scope.orientation);

        chrome.runtime.sendMessage({text:"createVerticalRuler"}, function(reponse){
        });
    }

    $scope.draw_ruler = function() {
        var minor_height = 3;
        var major_height = 5;
        var label_height = 10;

        var units = $scope.horizontal.units;

        if (units == "px" 
            || $scope.screen_diagonal_inches == 0 
            || $scope.screen_diagonal_inches === undefined) {
            var px_per_unit = window.getComputedStyle(document.getElementById(units)).width.slice(0, -2);
        } else {
            // We want to find a multipler that we can apply to the px_per_unit
            // so that it correctly represents the px_per_inch and px_per_cm
            // of the particular display in question. For a particular dimension
            // the way to do this is find the ratio between pixels and inches,
            // then use that as px_per_unit.
            var screen_diagonal_pixels = Math.sqrt(Math.pow(screen.width,2) +
                                                   Math.pow(screen.height,2));
            var ppi = screen_diagonal_pixels / $scope.screen_diagonal_inches;

            if (units == "cm") {
                var px_per_unit = ppi/cm_per_in;
            } else {
                var px_per_unit = ppi;
            }
        }
        
        console.log(px_per_unit + " px per " + units);

        var font_height = 12;
        var ctx = document.getElementById("ruler_canvas").getContext("2d");
        ctx.lineWidth = 1;
        ctx.font = String(font_height)+"px Arial";
        ctx.textAlign = "center";
        ctx.beginPath();

        var ppuMoveTo = function(x, y) {
            ctx.moveTo(x*px_per_unit-0.5, y);
        }
        var ppuLineTo = function(x, y) {
            ctx.lineTo(x*px_per_unit-0.5, y);
        }

        var spacing = $scope.horizontal.tick_spacing;
        var per_unit = $scope.horizontal.subdivisions_per_unit;

        for (var i=1; (i*spacing/per_unit*px_per_unit)<maximum_width; i++) {
            var x = i*spacing/per_unit;
            var y = 0;

            if (i % $scope.horizontal.minor_per_label == 0) {
                ppuMoveTo(x, y);
                ppuLineTo(x, y+label_height);
                ctx.fillText(x, x*px_per_unit, y+label_height+font_height);
            } else if (i % $scope.horizontal.minor_per_major == 0) {
                ppuMoveTo(x, y);
                ppuLineTo(x, y+major_height);
            } else {
                ppuMoveTo(x, y);
                ppuLineTo(x, y+minor_height);
            }
        }

        ctx.stroke();
    }

    angular.element(document).ready(function () {
        $scope.$watch('orientation', function () {
            console.log('init');
            tracker.sendAppView('Ruler-'+$scope.orientation);

            var canvas = document.getElementById("ruler_canvas");
            if ($scope.orientation == 'horizontal') {
                canvas.width = maximum_width
                canvas.height = document.body.clientHeight;
            } else {
                canvas.width = maximum_width
                canvas.height = document.body.clientWidth;
            }

            $scope.loadSettings();

            service.getConfig().addCallback(function(config) {
                var checkbox = document.getElementById('analytics');
                if (checkbox) {
                    checkbox.checked = config.isTrackingPermitted();
                    checkbox.onchange = function() {
                        config.setTrackingPermitted(checkbox.checked);
                    };
                }
            });

            var always_on_top_checkbox = document.getElementById('always_on_top');
            if (always_on_top_checkbox) {
                always_on_top_checkbox.onchange = function() {
                    chrome.app.window.current().setAlwaysOnTop(always_on_top_checkbox.checked);
                    $scope.saveSettings();
                }                
            }

            //document.getElementById("body").webkitRequestPointerLock();
            //document.addEventListener('webkitpointerlockchange', changeCallback, false);
        });
    });
}]);
